New OSLObject Interface

danieldresser-ie opened the pull request:
After a bunch of back and forth, hopefully this version of #2508 is getting fairly close to merging.

I still to read back through all the details myself, since it's been a while on some of this code, and I should probably add a bit more test coverage, but I thought I'd go ahead and put this up for you to take an initial look at.

danieldresser-ie commented:
Forgot to mention - there is a test failure, because as we discussed, this no longer is supporting shaders from before the shader serialization change.  It sounds like we're OK with that - does that mean I should just delete the test for really old backwards compatibility?

johnhaddon commented on src/GafferOSLUI/OSLObjectUI.cpp:
```diff
@@ -0,0 +1,271 @@
+				names.push_back( i->getName() );
+			}
+
+			std::string picked = nameMenuSignal()( "Connect To", names );
```
I'm confused. I thought we added `nameMenuSignal()` so you didn't need to build the proxy plugs, but now we're building the proxy plugs, then building a list of strings from them, then using the new menu, and then searching back in the proxies for the right name?

danieldresser-ie replied:
Gah, you're right.  I had originally thought that I would just prep a list of matching options as Data, and just do the createPlugFromData after picking one an option from the menu.  But I can't conveniently represent a Closure as Data, so it ended up keeping the proxyPlugs method I had orginally made to work around the lack of nameMenuSignal.  I guess the quickest solution is just to delete nameMenuSignal?

johnhaddon replied:
> I guess the quickest solution is just to delete nameMenuSignal?

Yeah, let's do that.

danieldresser-ie replied:
Conclusion from conversation with John:  the new signal is useful, whether we use it here or not.  I'll rename it from nameMenuSignal to just menuSignal, and include the commit, but reassess whether I actually use it in OSLObjectUI

danieldresser-ie commented on python/GafferOSLUI/OSLObjectUI.py:
```diff
@@ -41,6 +42,111 @@
+					label = newLabel
+				menuLabel = "/Custom/" + label[6:]
+			elif label == "multiInput":
+				menuLabel = "/Advanced/Multi Input"
```
My thought was that all of OutFloat and OutObject would be removed from the menus in the name of simplicity and clarity.  If for some reason you need to do some weird technical thing ( we haven't actually thought of any concrete use cases for the closure input yet ), I was thinking you can build a closure yourself using OSLCode.

Is that too obscure?  I really hesitate to put much emphasis on the possibility of closure inputs, since as far as I know all currently existing use cases would be better served by the new interface.

danieldresser-ie replied:
I should also say, in terms of the name, I can't really think of anything I like better, but I'm open to anything you come up with.  We just want something that says "that sounds weird and complicated, I shouldn't worry about it" to 95% of our users.

johnhaddon replied:
> My thought was that all of OutFloat and OutObject would be removed from the menus in the name of simplicity and clarity.

Can you add a commit for that then? Their continued existence meant that I assumed you intended them to be used with multi inputs.

> I was thinking you can build a closure yourself using OSLCode.

You definitely could, but it would be even less obvious than the OSLObject approach we're replacing. I guess this is all a bit theoretical without a solid use case, but I'm not certain that wanting to bundle several primvars in one connection is a sign that one also wants to learn about the low level details of ObjectProcessing.h and closures. I think there's a reasonable argument that if it's visible, it should be obvious what to do with it.

> I really hesitate to put much emphasis on the possibility of closure inputs, since as far as I know all currently existing use cases would be better served by the new interface.

I'm open to keeping them out of the menu until a need arises, at which point we can consider a friendlier name/workflow.

> I can't really think of anything I like better, but I'm open to anything you come up with.

"Variable Set" perhaps? For a decent workflow I think it needs to be a name that we can also use for an OSLObject-like shader that we intend folks to connect in. I'm not feeling confident about the naming though, so if you think just omitting the menu item is OK, then let's just do that.

johnhaddon replied:
Chatting offline, we agreed to rename the menu to "Advanced/Closure" and hide the Out* shaders. We're assuming that anyone who needs such an advanced feature will know a friendly local R&D person who can show them how to hook it up via OSLCode...

andrewkaufman commented:
>> does that mean I should just delete the test for really old backwards compatibility?

> Yes, as long as @andrewkaufman is happy

I'm not clear what we're referring to as really old backwards compatibility? Pre-ShaderNetwork? Or something even earlier than that? Which Gaffer version was this "really old" break made in and does IE have any user-published tools made before that? If so, how many (tip: @ivanimanishi might have a script for finding that I out)?

danieldresser-ie commented on python/GafferOSLUI/OSLObjectUI.py:
```diff
@@ -41,6 +42,111 @@
+		result = IECore.MenuDefinition()
+		usedNames = set()
+		for p in self.getPlug().children():
+			usedNames.add( p["name"].getValue() )
```
Thinking about it, we probably shouldn't count a name as already "used" based on a current expression value?  If I reject any plugs with non-null inputs in this loop, then the context shouldn't matter, right?

johnhaddon replied:
I'm not sure I agree - it doesn't matter where the name came from, if it's there, then it's in use. All an input tells us is that some other names could also be in use in other contexts. If you reject all plugs with non-null inputs, then you'll also be skipping promoted plugs, which might not be driven by an expression at all. [This](https://github.com/GafferHQ/gaffer/blob/master/src/Gaffer/Switch.cpp#L375-L376) is the reliable way to determine if something might be context varying - I've been thinking for a while it might be worth putting in PlugAlgo.

I guess it may not even be sufficient to scope our context if someone is using "scene:path" to drive the name though. I think that, and the avoidance of synchronous computes, is a stronger argument than anything about the name not being used if it comes from an expression.

danieldresser-ie replied:
Conclusion of conversation with John:  To avoid computes, I'll imitate the logic from switch for avoiding context dependend plugs.

danieldresser-ie commented on python/GafferOSLUI/OSLObjectUI.py:
```diff
@@ -41,6 +42,111 @@
+		for p in self.getPlug().children():
+			usedNames.add( p["name"].getValue() )
+
+		for label, constructor in sorted( _primitiveVariableNamesOptions.items() ):
```
Oh, yeah, switched this and hadn't switched the name yet.  Will do.

danieldresser-ie commented on python/GafferOSLUI/OSLObjectUI.py:
```diff
@@ -41,6 +42,111 @@
+
+		for label, constructor in sorted( _primitiveVariableNamesOptions.items() ):
+			if label.startswith( "custom" ):
+				if label in usedNames:
```
Yeah, makes sense.  This was feeling weird, but wasn't sure what to do to about it.  Your right that an exact match between the selected name and the name of the resulting plug is less important.

danieldresser-ie commented on src/GafferOSL/OSLObject.cpp:
```diff
@@ -385,3 +407,152 @@ ConstShadingEnginePtr OSLObject::shadingEngine( const Gaffer::Context *context )
+	}
+	catch( const std::exception &e )
+	{
+		// We call updateShader() from `plugSet()`
```
I haven't actually researched this heavily - I was going by the comment in OSLCode::updateShader ( and failed to update the comment when copying across ).  Do you think it's correct though, that code which can be triggered by childAddedSignal shouldn't throw exceptions?

johnhaddon replied:
I think the comment in OSLCode is probably right, and we should make `childAddedSignal()` robust to misbehaving slots rather than make the slots suppress things. The OSLCode case is a bit different to this one though I think - we expect to get exceptions as a matter of course in `OSLCode::updateShader()` due to the user entering invalid code, and we have a separate mechanism for reporting the error via `Node::errorSignal()`. Here on the other hand, any exception would be truly unexpected, and I'd probably want to know about it. Unless I missed something, I suggest we just remove the `try/catch`.

danieldresser-ie commented on src/GafferOSL/OSLObject.cpp:
```diff
@@ -96,17 +104,35 @@ bool initContextCompatibility()
+	ClosurePlugPtr codeOut = new ClosurePlug( "out", Plug::Out );
+	outputCombineOSLCode()->outPlug()->addChild( codeOut );
+
+	OSLShaderPtr closureToShaderConverter = new OSLShader( "__closureToShaderConverter" );
```
This would require OSLCode to output a "surface" instead of a "shader".  Perhaps this would be useful, but since this is an internal detail, I might prefer to leave it for later unless you're convinced it would be really easy to add?  How would you specify in the UI that you want an OSLCode that is a "surface"?

johnhaddon replied:
Whereabouts is the "surface" type required? OSL seems to be moving towards [removing shader types entirely](https://groups.google.com/d/msg/osl-dev/bVBZda-UsbI/L32I71UHBgAJ), so I wonder if we're making life harder than we need to somewhere.

danieldresser-ie commented on src/GafferOSL/OSLObject.cpp:
```diff
@@ -385,3 +407,152 @@ ConstShadingEnginePtr OSLObject::shadingEngine( const Gaffer::Context *context )
+				outputCombineOSLCode()->parametersPlug()->addChild( codeValuePlug );
+				if( (int)valueType == V2fPlugTypeId )
+				{
+					// For UVs, just connect up the first two components ( the third will be ignored by
```
I mostly just wanted something not confusing for the UI, and figured that since we accept V3f -> V2f connections anyway, I might as well use that as the plug type.

johnhaddon replied:
I can see why the V2f seems nice in the UI when looking at the OSLObject node in isolation, but I'm not convinced when you consider that its part of an ecosystem where we're not providing V2f anywhere...

danieldresser-ie commented:
I spoke with Andrew about the pre Gaffer-0.45 compatibility break.  I've republished all the utility boxes at IE that use OSLObject and hadn't been touched in the past year, so we should be OK with this break.

danieldresser-ie commented:
OK, I'm hoping to see the tests pass on Travis this time.  Still want to run through a more practical example interactively just to make sure everything is doing what it should in practice, but I think I've now addressed all feedback.  Would definitely be worth taking another look, since a bunch of things did need rejiggering.  But I think I'm now good to move onto applying this stuff to OSLImage.

johnhaddon commented:
I've restarted an OSX test that had an unrelated issue - can you merge if all runs smoothly?
